src/twisted/internet/asyncioreactor.py

```diff
@@ -9,42 +9,56 @@
 
 import errno
 
+from typing import Dict, Optional, Type, cast
+
 from zope.interface import implementer
 
 from twisted.logger import Logger
 from twisted.internet.posixbase import (PosixReactorBase, _NO_FILEDESC,
                                         _ContinuousPolling)
 from twisted.python.log import callWithLogger
 from twisted.python.runtime import seconds as runtimeSeconds
+from twisted.internet.abstract import FileDescriptor
 from twisted.internet.interfaces import IReactorFDSet
 
-try:
-    from asyncio import get_event_loop
-except ImportError:
-    raise ImportError("Requires asyncio.")
-
-# As per ImportError above, this module is never imported on python 2, but
-# pyflakes still runs on python 2, so let's tell it where the errors come from.
-from builtins import PermissionError, BrokenPipeError
+from asyncio import get_event_loop, AbstractEventLoop, SelectorEventLoop
 
 
 
 @implementer(IReactorFDSet)
 class AsyncioSelectorReactor(PosixReactorBase):
     """
     Reactor running on top of L{asyncio.SelectorEventLoop}.
+
+    On POSIX platforms, the default event loop is
+    L{asyncio.SelectorEventLoop}.
+    On Windows, the default event loop on Python 3.7 and older
+    is L{asyncio.WindowsSelectorEventLoop}, but on Python 3.8 and newer
+    the default event loop is L{asyncio.WindowsProactorEventLoop} which
+    is incompatible with L{AsyncioSelectorReactor}.
+    Applications that use L{AsyncioSelectorReactor} on Windows
+    with Python 3.8+ must call
+    C{asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())}
+    before instantiating and running L{AsyncioSelectorReactor}.
     """
     _asyncClosed = False
     _log = Logger()
 
-    def __init__(self, eventloop=None):
+    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
+        if not eventloop:
+            _eventloop = get_event_loop()  # type: AbstractEventLoop
+            eventloop = cast(Optional[SelectorEventLoop], _eventloop)
```

Comment:
If the cast is retained, it doesn't need the `Optional[]` does it?

```suggestion
            eventloop = cast(SelectorEventLoop, _eventloop)
```

Though I don't think the cast should be needed given the `isinstance()` check below.  Casts strike me as mostly just bypassing the type hinting and as such should be used sparingly.

```python
    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
        if not eventloop:
            selected_eventloop = get_event_loop()  # type: AbstractEventLoop
        else:
            selected_eventloop = eventloop

        # On Python 3.8+, asyncio.get_event_loop() on
        # Windows was changed to return a ProactorEventLoop
        if not isinstance(selected_eventloop, SelectorEventLoop):
            raise ValueError(
                "eventloop is: {} which is not a subclass of "
                "SelectorEventLoop".format(eventloop))

        self._asyncioEventloop = selected_eventloop  # type: SelectorEventLoop
```

(sorry this overlaps the other suggestions...)

Reply:
The cast and optional is still needed because since the`eventloop` parameter can be `None`, it needs to be `Optional` when specifying the type.

Reply:
Just casting to `SelectorEventLoop` seems to be fine given that that will satisfy an optional one.  Note that `eventloop` is not considered optional by mypy after the if.

https://mypy-play.net/?mypy=latest&python=3.8&gist=cb0b71f99dedf0745d49ce76a0f6dea3
```python
from asyncio import AbstractEventLoop, SelectorEventLoop, get_event_loop
from typing import Optional, cast

reveal_type(get_event_loop)

class C:
    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
        if eventloop is None:
            _eventloop = get_event_loop()  # type: AbstractEventLoop
            eventloop = cast(SelectorEventLoop, _eventloop)
        
        reveal_type(eventloop)
```
```
main.py:4: note: Revealed type is 'def () -> asyncio.events.AbstractEventLoop'
main.py:12: note: Revealed type is 'asyncio.unix_events._UnixSelectorEventLoop'
```

But given that we can comfortably skip the casting entirely and let mypy really check things properly, isn't that a better option?

Reply:
Here it is all together with both forms satisfying (at least a basic config of) mypy.

https://mypy-play.net/?mypy=latest&python=3.8&gist=112efdb08e33c25b8dbb6458261c574a
```python
from asyncio import AbstractEventLoop, SelectorEventLoop, get_event_loop
from typing import Optional, cast

reveal_type(get_event_loop)

class C:
    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
        if eventloop is None:
            _eventloop = get_event_loop()
            eventloop = cast(SelectorEventLoop, _eventloop)
        
        reveal_type(eventloop)

        if not isinstance(eventloop, SelectorEventLoop):
            raise ValueError(
                "eventloop is: {} which is not a subclass of "
                "SelectorEventLoop".format(eventloop))

        reveal_type(eventloop)
        self._asyncioEventloop = eventloop  # type: SelectorEventLoop


class D:
    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
        if eventloop is None:
            selected_eventloop = get_event_loop()
        else:
            selected_eventloop = eventloop

        reveal_type(selected_eventloop)

        if not isinstance(selected_eventloop, SelectorEventLoop):
            raise ValueError(
                "eventloop is: {} which is not a subclass of "
                "SelectorEventLoop".format(selected_eventloop))

        reveal_type(selected_eventloop)
        self._asyncioEventloop = selected_eventloop  # type: SelectorEventLoop
```

```
main.py:4: note: Revealed type is 'def () -> asyncio.events.AbstractEventLoop'
main.py:12: note: Revealed type is 'asyncio.unix_events._UnixSelectorEventLoop'
main.py:19: note: Revealed type is 'asyncio.unix_events._UnixSelectorEventLoop'
main.py:30: note: Revealed type is 'asyncio.events.AbstractEventLoop'
main.py:37: note: Revealed type is 'asyncio.unix_events._UnixSelectorEventLoop'
```

Reply:
Yes, mypy allows type refinement of a name in a scope

Reply:
@altendky I'm sorry, but I don't understand what you are asking for in your review feedback.  Can you clarify things more clearly?

Reply:
Do not use `cast()`.  It is basically disabling mypy.  Instead of letting mypy detect what the types are, you are telling mypy to just believe you that the types are what you want them to be.  There are presumably various corners cases where you find yourself between a bug in mypy and a not-yet-implemented feature in mypy and `cast()` is the 'easy' solution.  This isn't such a case.

The `cast()` is also fairly redundant with the `isinstance()` check which is another indicator that it isn't a good approach.

I provided alternative code without any casting that passes mypy and should end up doing the same thing as the code in the PR.  It also isn't particularly more complicated, if at all.

Reply:
OK, I incorporated your feedback and removed the `cast()`